src/main/scala/format/pgn/Parser.scala

```diff
@@ -98,296 +63,200 @@ object Parser {
+        "glyph"
+      )
 
-    def move: Parser[San] = castle | standard
+    val glyphs = glyph.rep0.map(gs => Glyphs.fromList(gs))
 
-    def castle =
-      (qCastle | kCastle) ~ suffixes ^^ { case side ~ suf =>
-        Castle(side) withSuffixes suf
-      }
-
-    val qCastle: Parser[Side] = ("O-O-O" | "o-o-o" | "0-0-0") ^^^ QueenSide
+    val x = P.char('x').?.map(_.isDefined)
 
-    val kCastle: Parser[Side] = ("O-O" | "o-o" | "0-0") ^^^ KingSide
+    val check = P.char('+').?.map(_.isDefined)
 
-    def standard: Parser[San] =
-      as("standard") {
-        (disambiguatedPawn | pawn | disambiguated | ambiguous | drop) ~ suffixes ^^ { case std ~ suf =>
-          std withSuffixes suf
-        }
-      }
+    val checkmate = (P.char('#') | P.string("++")).?.map(_.isDefined)
 
-    // e5
-    def pawn: Parser[Std] =
-      as("pawn") {
-        dest ^^ (de => Std(dest = de, role = Pawn))
-      }
+    val role = mapParser(Role.allByPgn, "role")
 
-    // Bg5
-    def ambiguous: Parser[Std] =
-      as("ambiguous") {
-        role ~ x ~ dest ^^ { case ro ~ ca ~ de =>
-          Std(dest = de, role = ro, capture = ca)
-        }
-      }
-
-    // B@g5
-    def drop: Parser[Drop] =
-      as("drop") {
-        role ~ "@" ~ dest ^^ { case ro ~ _ ~ po =>
-          Drop(role = ro, pos = po)
-        }
-      }
+    val file = mapParser(fileMap, "file")
 
-    // Bac3 Baxc3 B2c3 B2xc3 Ba2xc3
-    def disambiguated: Parser[Std] =
-      as("disambiguated") {
-        role ~ opt(file) ~ opt(rank) ~ x ~ dest ^^ { case ro ~ fi ~ ra ~ ca ~ de =>
-          Std(
-            dest = de,
-            role = ro,
-            capture = ca,
-            file = fi,
-            rank = ra
-          )
-        }
-      }
+    val rank = mapParser(rankMap, "rank")
 
-    // d7d5
-    def disambiguatedPawn: Parser[Std] =
-      as("disambiguated") {
-        opt(file) ~ opt(rank) ~ x ~ dest ^^ { case fi ~ ra ~ ca ~ de =>
-          Std(
-            dest = de,
-            role = Pawn,
-            capture = ca,
-            file = fi,
-            rank = ra
-          )
-        }
-      }
+    val dest: P[Pos] = mapParser(Pos.allKeys, "dest")
 
-    def suffixes: Parser[Suffixes] =
-      opt(promotion) ~ checkmate ~ check ~ glyphs ^^ { case p ~ cm ~ c ~ g =>
-        Suffixes(c, cm, p, g)
-      }
+    val promotable = Role.allPromotableByPgn mapKeys (_.toUpper)
 
-    def glyphs: Parser[Glyphs] =
-      as("glyphs") {
-        rep(glyph) ^^ Glyphs.fromList
-      }
+    val promotion: P[PromotableRole] = P.char('=').?.with1 *> mapParser(promotable, "promotion")
 
-    def glyph: Parser[Glyph] =
-      as("glyph") {
-        mapParser(
-          Glyph.MoveAssessment.all.sortBy(_.symbol.length).map { g =>
-            g.symbol -> g
-          },
-          "glyph"
-        )
-      }
+    // e5
+    val pawn: P[Std] = dest.map(Std(_, Pawn))
 
-    val x = exists("x")
+    // Bg5
+    val ambigous: P[Std] = (role ~ x ~ dest).map { case ((ro, ca), de) =>
```

Comment:
fyi, with `cats.implicits._` or just the syntax import you could also do:

```scala
(role, x, dest).mapN(Std(_, _, _))
```

Reply:
Awesome, I didn't know this.